Only Retry POST Receipt Paths for 429

fire-at-will opened the pull request:
### Description
This PR modifies the HTTPClient's retry logic to only perform retries on requests to the `/v1/receipts` URL path that result in a 429 status code.

#### Testing
- Updated existing tests to set the `isRetryable` field correctly
- Added unit tests to `HTTPClient.Request` to ensure that the `isRetryable` field is being set properly
- Wrote new unit & integration tests to ensure that we don't retry requests to unsupported endpoints (like `/v1/subscribers/identify`)

fire-at-will commented on Sources/Networking/HTTPClient/HTTPClient.swift:
```diff
@@ -666,17 +668,15 @@ extension HTTPClient {
-            isRetryable = Bool(isRetryableString.lowercased()) ?? true
-        } else {
-            isRetryable = true
+        let doesServerAllowRetryString = urlResponse.value(forHTTPHeaderField: ResponseHeader.isRetryable.rawValue)
```
No change in the logic, just renaming `isRetryable` to `doesServerAllowRetry` to be more clear

fire-at-will commented:
@RCGitBot please test

fire-at-will commented:
Some of the test checks look a little funny right now because it took me a moment to figure out how to run all of the tests, I'll make sure they're all good to go before merging 💪

fire-at-will commented on Tests/BackendIntegrationTests/OtherIntegrationTests.swift:
```diff
@@ -173,7 +175,7 @@ class OtherIntegrationTests: BaseBackendIntegrationTests {
 
         let result = try await self.purchases.productEntitlementMapping().entitlementsByProduct
-        expect(result).to(haveCount(18))
+        expect(result).to(haveCount(21))
```
I'm not sure when (the audit logs aren't loading for me), but we now have 21 iOS products in the StoreKit Integration Tests app instead of 18

fire-at-will replied:
I'm looking into this a little more to confirm why this has changed

fire-at-will replied:
The most recently added product to the project was on June 25, so this isn't an issue with a new product being present. Continuing to dig 👀

fire-at-will replied:
The test is currently failing in main with the value `18`, but passed during yesterday's release pipeline 🧐


https://app.circleci.com/pipelines/github/RevenueCat/purchases-ios/20914/workflows/68cd311d-9aec-47a8-8f55-889bef8bdabc

aboedo replied:
@RevenueCat/coresdk thoughts?

fire-at-will replied:
Josh and I took a look at some network logs and found that the backend was returning 18 mappings on Tuesday when the test passed, but is now returning 21 in the network calls to `/v1/product_entitlement_mapping`. This is happening on this branch, main, and the 4.32.2 release. We now believe that 21 is the correct number and are following up with the backend team to investigate what caused the change to take so long to appear.

aboedo commented on Sources/Networking/HTTPClient/HTTPClient.swift:
```diff
@@ -666,17 +668,15 @@ extension HTTPClient {
-        } else {
-            isRetryable = true
+        let doesServerAllowRetryString = urlResponse.value(forHTTPHeaderField: ResponseHeader.isRetryable.rawValue)
+        var doesServerAllowRetry: Bool = true
+        if let doesServerAllowRetryString = doesServerAllowRetryString {
+            doesServerAllowRetry = Bool(doesServerAllowRetryString.lowercased()) ?? true
```
I'm sure this version is more clever, but it was kinda clearer before the ternary IMO

fire-at-will replied:
I'm okay reverting it! I'll keep the name `doesServerAllowRetry` over `isRetryable` though 😅

aboedo commented on Sources/Networking/HTTPClient/HTTPRequest.swift:
```diff
@@ -22,26 +22,49 @@ struct HTTPRequest {
-
-    init(method: Method, path: HTTPRequest.Path, nonce: Data? = nil) {
-        self.init(method: method, requestPath: path, nonce: nonce)
+    /// Whether or not this reqeust should be retried by the HTTPClient for certain status codes.
```
typo `request`

aboedo commented on Tests/BackendIntegrationTests/StoreKitIntegrationTests.swift:
```diff
@@ -935,6 +935,7 @@ class StoreKit1IntegrationTests: BaseStoreKitIntegrationTests {
     }
 
     func testVerifyPurchaseGrantsEntitlementsThroughOnRetryAfter429() async throws {
+        defer { HTTPStubs.removeAllStubs() }
```
should this just be a setup or teardown?

fire-at-will replied:
Moved it to the `tearDown` functions 😎

aboedo reviewed (commented):
looking good so far

fire-at-will commented:
Hey @aboedo! Since your last review, I've made the following changes:

### Code Changes
- Reverted the syntax when computing `doesServerAllowRetry`
- Fixed the "request" typo in `HTTPRequest.swift`
- Moved the `HTTPStubs.removeAllStubs()` calls in the tests from a `defer` statements to the `tearDown` functions

### Test Fixes/Clarifications
- Fixed the configuration issues that were causing the `CustomEntitlementsComputationIntegrationTests` to fail
- Dug into why `OtherIntegrationTests.testProductEntitlementMapping` was returning 21 mappings instead of 18. Josh and I found that Khepri was returning 18 mappings on Tuesday when the test passed, but is now returning 21 in the network calls to `/v1/product_entitlement_mapping`. We now believe that 21 is the correct number and are following up with the backend team to investigate what caused the change to take so long to appear.

This should be ready for another look!
